src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,415 @@
+            return False
+
+        # check the convergence conditions of the line search
+        # the first condition corresponds to the usual armijo descent criterion
+        # the second condition checks if the new relative error is still within the trust region
+        output_error = surrogate.estimate_output_error(new_mu)
+        return (current_value <= initial_value - (armijo_alpha / step) * np.linalg.norm(new_mu - initial_mu)**2
+            and output_error / abs(current_value) <= radius)
+
+    data = {'subproblem_data': []}
+
+    # compute norms
+    mu_norm = np.linalg.norm(mu)
+    update_norms = []
+    foc_norms = []
+    data['mus'] = [mu.copy()]
+
+    old_rom_output = surrogate.output(mu)
+    old_fom_output = fom.output(mu)
+
+    first_order_criticality = np.inf
+    iteration = 0
+    while True:
+        with logger.block(f'Starting adaptive TR algorithm iteration {iteration + 1} with radius {radius}...'):
+            rejected = False
+
+            if iteration >= miniter:
+                if first_order_criticality < tol_criticality:
+                    logger.info(
+                        f'TR converged in {iteration} iterations because first order criticality tolerance of' \
+                        f' {tol_criticality} was reached. The reduced basis is of size {surrogate.rb_size()}.')
+                    break
+                if iteration >= maxiter:
+                    logger.info(f'Maximum iterations reached. Failed to converge after {iteration} iterations.')
+                    raise TRError('Failed to converge after the maximum amount of iterations.')
+
+            iteration += 1
+
+            # solve the subproblem using bfgs
+            old_mu = mu.copy()
+
+            with logger.block(f'Solving subproblem for mu {mu} with BFGS...'):
+                mu, sub_data = error_aware_bfgs(
+                    surrogate, parameter_space, initial_guess=mu, miniter=miniter_subproblem,
+                    maxiter=maxiter_subproblem, rtol_output=rtol_output, rtol_mu=rtol_mu, tol_sub=tol_sub,
+                    line_search_params=line_search_params, stagnation_window=stagnation_window,
+                    stagnation_threshold=stagnation_threshold, error_aware=True,
+                    error_criterion=error_aware_bfgs_criterion,
+                    line_search_error_criterion=error_aware_line_search_criterion)
+
+            # first BFGS iterate is AGC point
+            index = 1 if len(sub_data['mus']) > 1 else 0
+            compare_output = surrogate.output(sub_data['mus'][index])
+            estimate_output = surrogate.estimate_output_error(mu)
+            current_output = surrogate.output(mu)
+
+            with logger.block('Running output checks for TR parameters.'):
+                if current_output + estimate_output < compare_output:
+                    surrogate.extend(mu)
+                    current_fom_output = fom.output(mu)
```

Comment:
This fom evaluation is not couted, right?

Reply:
added remarks that this solution is available by using caching.